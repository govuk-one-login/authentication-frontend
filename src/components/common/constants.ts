import { PATH_NAMES } from "../../app.constants";
import { getNextState } from "./state-machine/state-machine";

export const SECURITY_CODE_ERROR = "actionType";

export enum SecurityCodeErrorType {
  MfaMaxCodesSent = "mfaMaxCodesSent",
  MfaBlocked = "mfaBlocked",
  MfaMaxRetries = "mfaMaxRetries",
  OtpMaxCodesSent = "otpMaxCodesSent",
  OtpBlocked = "otpBlocked",
  OtpMaxRetries = "otpMaxRetries",
  EmailMaxCodesSent = "emailMaxCodesSent",
  EmailBlocked = "emailBlocked",
  EmailMaxRetries = "emailMaxRetries",
  AuthAppMfaNoCodeValidator = "authAppMfaNoCodeValidator",
  AuthAppMfaMaxRetries = "authAppMfaMaxRetries",
  ChangeSecurityCodesEmailMaxCodesSent = "changeSecurityCodesEmailMaxCodesSent",
  ChangeSecurityCodesEmailBlocked = "changeSecurityCodesEmailBlocked",
  ChangeSecurityCodesEmailMaxRetries = "changeSecurityCodesEmailMaxRetries",
  InvalidPasswordResetCodeMaxRetries = "invalidPasswordResetCodeMaxRetries",
}

export const ERROR_CODES = {
  SESSION_ID_MISSING_OR_INVALID: 1000,
  RESET_PASSWORD_LINK_MAX_RETRIES_REACHED: 1022,
  RESET_PASSWORD_LINK_BLOCKED: 1023,
  NEW_PASSWORD_SAME_AS_EXISTING: 1024,
  ENTERED_INVALID_MFA_MAX_TIMES: 1027,
  VERIFY_EMAIL_CODE_REQUEST_BLOCKED: 1029,
  VERIFY_PHONE_NUMBER_CODE_REQUEST_BLOCKED: 1030,
  VERIFY_PHONE_NUMBER_MAX_CODES_SENT: 1032,
  VERIFY_EMAIL_MAX_CODES_SENT: 1031,
  ENTERED_INVALID_VERIFY_EMAIL_CODE_MAX_TIMES: 1033,
  ENTERED_INVALID_VERIFY_PHONE_NUMBER_CODE_MAX_TIMES: 1034,
  MFA_SMS_MAX_CODES_SENT: 1025,
  MFA_CODE_REQUESTS_BLOCKED: 1026,
  INVALID_MFA_CODE: 1035,
  INVALID_VERIFY_EMAIL_CODE: 1036,
  INVALID_VERIFY_PHONE_NUMBER_CODE: 1037,
  ENTERED_INVALID_PASSWORD_RESET_CODE_MAX_TIMES: 1039,
  INVALID_PASSWORD_MAX_ATTEMPTS_REACHED: 1028,
  RESET_PASSWORD_INVALID_CODE: 1021,
  PASSWORD_IS_COMMON: 1040,
  AUTH_APP_INVALID_CODE_MAX_ATTEMPTS_REACHED: 1042,
  AUTH_APP_INVALID_CODE: 1043,
  ACCOUNT_LOCKED: 1045,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_MAX_CODES_SENT: 1046,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_CODE_REQUEST_BLOCKED: 1047,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_INVALID_CODE: 1048,
};

export const ERROR_CODE_MAPPING: { [p: string]: string } = {
  [ERROR_CODES.ACCOUNT_LOCKED]: pathWithQueryParam(
    PATH_NAMES["ACCOUNT_LOCKED"]
  ),
  [ERROR_CODES.INVALID_PASSWORD_MAX_ATTEMPTS_REACHED]: pathWithQueryParam(
    PATH_NAMES["ACCOUNT_LOCKED"]
  ),
  [ERROR_CODES.MFA_SMS_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaMaxCodesSent
  ),
  [ERROR_CODES.MFA_CODE_REQUESTS_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_MFA_MAX_TIMES]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaMaxRetries
  ),
  [ERROR_CODES.VERIFY_PHONE_NUMBER_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.OtpMaxCodesSent
  ),
  [ERROR_CODES.VERIFY_PHONE_NUMBER_CODE_REQUEST_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.OtpBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_VERIFY_PHONE_NUMBER_CODE_MAX_TIMES]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.OtpMaxRetries
    ),
  [ERROR_CODES.VERIFY_EMAIL_CODE_REQUEST_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_VERIFY_EMAIL_CODE_MAX_TIMES]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailMaxRetries
  ),
  [ERROR_CODES.ENTERED_INVALID_PASSWORD_RESET_CODE_MAX_TIMES]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.InvalidPasswordResetCodeMaxRetries
    ),
  [ERROR_CODES.VERIFY_EMAIL_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailMaxCodesSent
  ),
  [ERROR_CODES.AUTH_APP_INVALID_CODE_MAX_ATTEMPTS_REACHED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.AuthAppMfaMaxRetries
  ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_MAX_CODES_SENT]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailMaxCodesSent
    ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_INVALID_CODE]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailMaxRetries
    ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_CODE_REQUEST_BLOCKED]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_WAIT"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailBlocked
    ),
};

export function getErrorPathByCode(errorCode: number): string | undefined {
  const nextPath = ERROR_CODE_MAPPING[errorCode.toString()];

  if (!nextPath) {
    return undefined;
  }

  return nextPath;
}

export function pathWithQueryParam(
  path: string,
  queryParam?: string,
  value?: string | SecurityCodeErrorType
): string {
  if (queryParam && value) {
    const queryParams = new URLSearchParams({
      [queryParam]: value,
    }).toString();

    return path + "?" + queryParams;
  }

  return path;
}

export function getNextPathAndUpdateJourney(
  req: any,
  path: string,
  event: string,
  ctx?: any,
  sessionId?: string
): string {
  const nextState = getNextState(path, event, ctx);

  req.session.user.journey = {
    nextPath: nextState.value,
    optionalPaths:
      Object.keys(nextState.meta).length > 0
        ? nextState.meta["AUTH." + nextState.value].optionalPaths
        : [],
  };

  req.log.info(
    `User journey transitioned from ${req.path} to ${nextState.value} with session id ${sessionId}`
  );

  if (!nextState) {
    throw Error(
      `Invalid user journey. No transition found from ${path} with event ${event}`
    );
  }

  return nextState.value;
}

export const JOURNEY_TYPE = {
  SIGN_IN: "sign-in",
  CREATE_ACCOUNT: "create-account",
};
