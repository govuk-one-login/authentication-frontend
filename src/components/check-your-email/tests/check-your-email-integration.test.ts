import { describe } from "mocha";
import { expect, sinon, request } from "../../../../test/utils/test-utils.js";
import nock from "nock";
import * as cheerio from "cheerio";
import decache from "decache";
import {
  API_ENDPOINTS,
  HTTP_STATUS_CODES,
  PATH_NAMES,
} from "../../../app.constants.js";
import { ERROR_CODES, SecurityCodeErrorType } from "../../common/constants.js";
import type { NextFunction, Request, Response } from "express";
import { getPermittedJourneyForPath } from "../../../../test/helpers/session-helper.js";
describe("Integration:: check your email", () => {
  let token: string | string[];
  let cookies: string;
  let app: any;
  let baseApi: string;

  before(async () => {
    process.env.SUPPORT_CHECK_EMAIL_FRAUD = "1";
    decache("../../../app");
    decache("../../../middleware/session-middleware");
    const sessionMiddleware = await import(
      "../../../middleware/session-middleware"
    );

    sinon
      .stub(sessionMiddleware, "validateSessionMiddleware")
      .callsFake(function (
        req: Request,
        res: Response,
        next: NextFunction
      ): void {
        res.locals.sessionId = "tDy103saszhcxbQq0-mjdzU854";

        req.session.user = {
          email: "test@test.com",
          journey: getPermittedJourneyForPath(PATH_NAMES.CHECK_YOUR_EMAIL),
        };

        next();
      });

    app = await (await import("../../../app")).createApp();
    baseApi = process.env.FRONTEND_API_BASE_URL;

    await request(app, (test) => test.get(PATH_NAMES.CHECK_YOUR_EMAIL), {
      expectAnalyticsPropertiesMatchSnapshot: false,
    }).then((res) => {
      const $ = cheerio.load(res.text);
      token = $("[name=_csrf]").val();
      cookies = res.headers["set-cookie"];
    });
  });

  beforeEach(() => {
    nock.cleanAll();
  });

  after(() => {
    sinon.restore();
    app = undefined;
    delete process.env.SUPPORT_CHECK_EMAIL_FRAUD;
  });

  it("should return verify email page", async () => {
    await request(app, (test) =>
      test.get(PATH_NAMES.CHECK_YOUR_EMAIL).expect(200)
    );
  });

  it("should return error when csrf not present", async () => {
    await request(app, (test) =>
      test
        .post(PATH_NAMES.CHECK_YOUR_EMAIL)
        .type("form")
        .send({
          code: "123456",
        })
        .expect(403)
    );
  });

  it("should return validation error when code not entered", async () => {
    await request(app, (test) =>
      test
        .post(PATH_NAMES.CHECK_YOUR_EMAIL)
        .type("form")
        .set("Cookie", cookies)
        .send({
          _csrf: token,
          code: "",
        })
        .expect(function (res) {
          const $ = cheerio.load(res.text);
          expect($("#code-error").text()).to.contains("Enter the code");
        })
        .expect(400)
    );
  });

  it("should return validation error when code is less than 6 characters", async () => {
    await request(app, (test) =>
      test
        .post(PATH_NAMES.CHECK_YOUR_EMAIL)
        .type("form")
        .set("Cookie", cookies)
        .send({
          _csrf: token,
          code: "2",
        })
        .expect(function (res) {
          const $ = cheerio.load(res.text);
          expect($("#code-error").text()).to.contains(
            "Enter the code using only 6 digits"
          );
        })
        .expect(400)
    );
  });

  it("should return validation error when code is greater than 6 characters", async () => {
    await request(app, (test) =>
      test
        .post(PATH_NAMES.CHECK_YOUR_EMAIL)
        .type("form")
        .set("Cookie", cookies)
        .send({
          _csrf: token,
          code: "1234567",
        })
        .expect(function (res) {
          const $ = cheerio.load(res.text);
          expect($("#code-error").text()).to.contains(
            "Enter the code using only 6 digits"
          );
        })
        .expect(400)
    );
  });

  it("should return validation error when code entered contains letters", async () => {
    await request(app, (test) =>
      test
        .post(PATH_NAMES.CHECK_YOUR_EMAIL)
        .type("form")
        .set("Cookie", cookies)
        .send({
          _csrf: token,
          code: "12ert-",
        })
        .expect(function (res) {
          const $ = cheerio.load(res.text);
          expect($("#code-error").text()).to.contains(
            "Enter the code using only 6 digits"
          );
        })
        .expect(400)
    );
  });

  it("should redirect to /create-password when valid code entered", async () => {
    nock(baseApi)
      .post(API_ENDPOINTS.VERIFY_CODE)
      .once()
      .reply(HTTP_STATUS_CODES.NO_CONTENT, {});

    nock(baseApi)
      .post(API_ENDPOINTS.CHECK_EMAIL_FRAUD_BLOCK)
      .once()
      .reply(HTTP_STATUS_CODES.OK, {
        email: "test@test.com",
        isBlockedStatus: "Pending",
      });

    await request(app, (test) =>
      test
        .post(PATH_NAMES.CHECK_YOUR_EMAIL)
        .type("form")
        .set("Cookie", cookies)
        .send({
          _csrf: token,
          code: "123456",
        })
        .expect("Location", PATH_NAMES.CREATE_ACCOUNT_SET_PASSWORD)
        .expect(302)
    );
  });

  it("should return validation error when incorrect code entered", async () => {
    nock(baseApi).post(API_ENDPOINTS.VERIFY_CODE).once().reply(400, {
      code: ERROR_CODES.INVALID_VERIFY_EMAIL_CODE,
      success: false,
    });

    nock(baseApi)
      .post(API_ENDPOINTS.CHECK_EMAIL_FRAUD_BLOCK)
      .once()
      .reply(HTTP_STATUS_CODES.OK, {
        email: "test@test.com",
        isBlockedStatus: "Pending",
      });

    await request(app, (test) =>
      test
        .post(PATH_NAMES.CHECK_YOUR_EMAIL)
        .type("form")
        .set("Cookie", cookies)
        .send({
          _csrf: token,
          code: "123455",
        })
        .expect(function (res) {
          const $ = cheerio.load(res.text);
          expect($("#code-error").text()).to.contains(
            "The code you entered is not correct, or may have expired, try entering it again or request a new code"
          );
        })
        .expect(400)
    );
  });

  it("should return error page when incorrect code entered more than 5 times", async () => {
    nock(baseApi).post(API_ENDPOINTS.VERIFY_CODE).times(6).reply(400, {
      code: ERROR_CODES.ENTERED_INVALID_VERIFY_EMAIL_CODE_MAX_TIMES,
      success: false,
    });

    nock(baseApi)
      .post(API_ENDPOINTS.CHECK_EMAIL_FRAUD_BLOCK)
      .once()
      .reply(HTTP_STATUS_CODES.OK, {
        email: "test@test.com",
        isBlockedStatus: "Pending",
      });

    await request(app, (test) =>
      test
        .post(PATH_NAMES.CHECK_YOUR_EMAIL)
        .type("form")
        .set("Cookie", cookies)
        .send({
          _csrf: token,
          code: "123455",
        })
        .expect(
          "Location",
          `${PATH_NAMES.SECURITY_CODE_INVALID}?actionType=${SecurityCodeErrorType.EmailMaxRetries}`
        )
        .expect(302)
    );
  });
});
