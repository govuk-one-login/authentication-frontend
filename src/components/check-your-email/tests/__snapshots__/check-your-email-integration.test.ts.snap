// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration:: check your email should return error page when incorrect code entered more than 5 times 1`] = `
Object {
  "contentId": "1fef9388-34cd-4ea2-b899-a66b7327d2f7",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
}
`;

exports[`Integration:: check your email should return error when csrf not present 1`] = `
Object {
  "contentId": undefined,
  "taxonomyLevel1": undefined,
  "taxonomyLevel2": undefined,
}
`;

exports[`Integration:: check your email should return validation error when code entered contains letters 1`] = `
Object {
  "contentId": "054e1ea8-97a8-461a-a964-07345c80098e",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
}
`;

exports[`Integration:: check your email should return validation error when code is greater than 6 characters 1`] = `
Object {
  "contentId": "054e1ea8-97a8-461a-a964-07345c80098e",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
}
`;

exports[`Integration:: check your email should return validation error when code is less than 6 characters 1`] = `
Object {
  "contentId": "054e1ea8-97a8-461a-a964-07345c80098e",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
}
`;

exports[`Integration:: check your email should return validation error when code not entered 1`] = `
Object {
  "contentId": "054e1ea8-97a8-461a-a964-07345c80098e",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
}
`;

exports[`Integration:: check your email should return validation error when incorrect code entered 1`] = `
Object {
  "contentId": "054e1ea8-97a8-461a-a964-07345c80098e",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
}
`;

exports[`Integration:: check your email should return verify email page 1`] = `
Object {
  "contentId": "054e1ea8-97a8-461a-a964-07345c80098e",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "create account",
}
`;
